tests/transpiler/qasm/test_qasm_to_cudaq.py

```diff
@@ -0,0 +1,329 @@
+# Copyright (C) 2024 qBraid
+#
+# This file is part of the qBraid-SDK
+#
+# The qBraid-SDK is free software released under the GNU General Public License v3
+# or later. You can redistribute and/or modify it under the terms of the GPL v3.
+# See the LICENSE file in the project root or <https://www.gnu.org/licenses/gpl-3.0.html>.
+#
+# THERE IS NO WARRANTY for the qBraid-SDK, as per Section 15 of the GPL v3.
+
+"""
+Unit tests for OpenQASM 3.0 to CUDA-Q kernel transpilation.
+
+"""
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import numpy as np
+import pytest
+from openqasm3.parser import parse
+from qiskit.circuit.random import random_clifford_circuit
+from qiskit.qasm3 import dumps as qasm3_dumps
+from qiskit.qasm3 import loads as qasm3_loads
+from qiskit_aer import StatevectorSimulator
+
+from qbraid.interface import assert_allclose_up_to_global_phase, circuits_allclose
+from qbraid.transpiler.conversions.openqasm3 import openqasm3_to_cudaq
+from qbraid.transpiler.conversions.qasm2.qasm2_to_qasm3 import qasm2_to_qasm3
+
+cudaq = pytest.importorskip("cudaq")
+
+if TYPE_CHECKING:
+    from cudaq import PyKernel
+    from qiskit import QuantumCircuit
+
+
+def _check_output(qasm3_str_in: str, cudaq_out: PyKernel, atol=1e-7, method="circ"):
+    """Check output given the input OpenQASM 3.0 program and the CUDA-Q kernel.
+    Args
+        qasm3_str_in (str): OpenQASM 3.0 input program
+        cudaq_out (PyKernel): CUDA-Q kernel equivalent to input string
+        method (str):
+        - 'circ' to use CUDA-Q translate -> OpenQASM 2 -> Qiskit circuits -> all close.
+        - 'state' to compute statevectors -> all close. requires no measurement.
+    """
+    if method == "circ":
+        qasm2_str_out = cudaq.translate(cudaq_out, format="openqasm2")
+        qasm3_str_out = qasm2_to_qasm3(qasm2_str_out)
+        circ_in, circ_out = qasm3_loads(qasm3_str_in), qasm3_loads(qasm3_str_out)
+
+        assert circuits_allclose(circ_in, circ_out, atol=atol)
+    elif method == "state":
+        circ_in: QuantumCircuit = qasm3_loads(qasm3_str_in)
+        state_out = np.array(cudaq.get_state(cudaq_out))
+
+        sim = StatevectorSimulator()
+        job = sim.run(circ_in, shots=1)
+        res = job.result().results[0]
+        state_in = res.data.statevector.data
+
+        assert_allclose_up_to_global_phase(state_in, state_out, atol=atol)
+
+
+def test_openqasm3_to_cudaq():
+    """Test converting an OpenQASM3 program to a CUDA-Q kernel."""
+
+    qasm3_str_in = """
+    OPENQASM 3.0;
+    include "stdgates.inc";
+
+    qubit[1] q;
+    bit[1] b;
+
+    h q[0];
+    x q[0];
+    y q[0];
+    z q[0];
+    t q[0];
+    s q[0];
+
+    b[0] = measure q[0];
+    """
+    qasm3_in = parse(qasm3_str_in)
```

Comment:
If we are allowing `str` in the `openqasm3_to_cudaq`, you can directly pass in the qasm string. 
Same comment applies for all tests.